sql/temporary_tables.cc

```diff
@@ -795,7 +795,8 @@ void THD::mark_tmp_table_as_free_for_reuse(TABLE *table)
 
   table->query_id= 0;
   table->file->ha_reset();
-
+  if (unlikely(table->needs_reopen()) && table->s->table_type == TABLE_TYPE_SEQUENCE)
```

Comment:
Without checking for sequences we get the failure
```
+++ /home/anel/GitHub/mariadb/server/src/11.2/mysql-test/main/alter_table_errors.reject	2023-08-06 10:38:03.508451890 +0200
@@ -15,7 +15,7 @@
 Table	Create Table
 t1	CREATE TEMPORARY TABLE `t1` (
   `a` int(11) DEFAULT NULL,
-  `v` int(11) GENERATED ALWAYS AS (`a`) VIRTUAL
+  `v` int(11) GENERATED ALWAYS AS (`?`) VIRTUAL
 ) ENGINE=MyISAM DEFAULT CHARSET=latin1 COLLATE=latin1_swedish_ci
 create temporary table t2 (a int, v int as (a));
 lock table t2 write;
@@ -25,7 +25,7 @@
 Table	Create Table
 t2	CREATE TEMPORARY TABLE `t2` (
   `a` int(11) DEFAULT NULL,
-  `v` int(11) GENERATED ALWAYS AS (`a`) VIRTUAL
+  `v` int(11) GENERATED ALWAYS AS (`?`) VIRTUAL
 ) ENGINE=MyISAM DEFAULT CHARSET=latin1 COLLATE=latin1_swedish_ci
 drop temporary table t1, t2;
 create temporary table t1 (a int);
```
and 
```
--- /home/anel/GitHub/mariadb/server/src/11.2/mysql-test/main/temp_table.result2023-08-02 16:36:05.578615581 +0200
+++ /home/anel/GitHub/mariadb/server/src/11.2/mysql-test/main/temp_table.reject2023-08-06 10:38:02.664453558 +0200
@@ -595,8 +595,8 @@
 UPDATE t1 SET b = 3;
 SELECT * FROM t1;
 a	b
-2001-01-01 10:20:30	3
-2001-01-01 10:20:30	3
+NULL	3
+NULL	3
 DROP TEMPORARY TABLE t1;
```
Is this related to the [MDEV-31853](https://jira.mariadb.org/browse/MDEV-31853) ?